### PR TITLE
Making Contact immutable and Adding a copyWith() method.

### PR DESCRIPTION
**Pull request by yassinsameh**
A copyWith() method can be useful for cloning an Object or getting a copy while changing one or more fields. I also found it useful when doing Load testing, i needed to have thousands of unique contacts and having the copyWith method was helpful.

### CODE REVIEW

**Comment by yassinsameh** [comment]
Commits include tests added and previous examples/test updated for the new syntax. Was having edge case bugs because of contacts being mutable, think this change makes it much cleaner and makes projects less error prone, please take a look to validate. Great job overall on the package. @joachimvalente

File: lib/contact.dart [hunk 1]
```diff
@@ -220,6 +232,52 @@ class Contact {
...
+    bool? thumbnailFetched,
+    bool? photoFetched,
+    bool? isUnified,
+    bool? propertiesFetched,
+  }) {
+    return Contact(
+      id: id ?? this.id,
+      displayName: displayName ?? this.displayName,
+      thumbnail: thumbnail ?? this.thumbnail,
+      photo: photo ?? this.photo,
```
Comment (jjochen) [inline lib/contact.dart:261]: this does not work if you want to set the photo or thumbnail to `null`.
```dart
final copy = contact.copyWith(photo: null);
```
If the original photo wasn't `null`, this implementation will not set it to `null`.

For this to work, you need to some kind of wrapper around the actual value. 
You can use a Function for that. Something like this:
```dart
  Contact copyWith({
    Uint8List? Function()? photo,
  }) {
    return Contact(
      photo: photo != null ? photo() : this.photo,
    );
  }
```

File: test/contact_test.dart [hunk 2]
```diff
@@ -22,16 +22,31 @@ void main() {
...
+    expect(contact1.displayName == contact2.displayName, false);
+  });
+  test(
+      'When using copyWith() with attributes, validate equality '
+      'operator function detects objects difference', () {
+    final Contact contact1 = Contact(id: "1", displayName: "x");
+    final Contact contact2 = contact1.copyWith(id: "2");
+
+    expect(contact1 == contact2, false);
+  });
```
Comment (jjochen) [inline test/contact_test.dart:49]: You should add tests for copy with `null`. Something like
```dart
expect(contactWithPhoto.copyWith(photo: null).photo, isNull);
```
Currently this would fail.

**Review by jjochen — CHANGES_REQUESTED** [review]

File: lib/contact.dart (hunk 1 above)
Comment (yassinsameh) [reply lib/contact.dart]: Yes well spotted @jjochen , i think following Remi's answer [here](https://stackoverflow.com/questions/63928513/differentiate-undefined-and-null-in-dart) which the package "freezed" uses under the hood would be best: , just commited that with tests.

Though it might be a good idea to make all classes use freezed instead for a more robust management, and override any methods when needed, it's quite a bit of a big change that i am unable to take on at the moment, so perhaps that would be be on the plan for future development.

**Comment by dphans** [comment]
Follow Clean Architecture, I think you no need to do that, instead, custom domain Model (eg. ContactModel) and custom `copyWith` as you need.
